src/compiler/crystal/config.cr

```diff
@@ -34,8 +34,41 @@ module Crystal
       Time.unix(time).to_s("%Y-%m-%d")
     end
 
-    def self.default_target_triple
-      {{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple
+    @@default_target : Crystal::Codegen::Target?
+
+    def self.default_target : Crystal::Codegen::Target
+      @@default_target ||= begin
+        target = Crystal::Codegen::Target.new({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple)
+
+        if target.linux?
+          # The linux binary runs as well on linux-gnu as linux-musl, but the
+          # default target needs to match the C library used on the current
+          # system.
+          # This can be automatically detected from the output of `ldd --version`
+          # in order to use the appropriate environment target.
+          default_libc = target.gnu? ? "-gnu" : "-musl"
+
+          target = Crystal::Codegen::Target.new(target.to_s.sub(default_libc, "-#{runtime_libc}"))
```

Comment:
why not make setters for `architecture`, `vendor` and `environment`.  You would not have to make new targets and you would not have to rely on a `.sub` for the replacement, it would be an exact replacement.

Reply:
Yeah, that would actually be more complicated because `environment` is an array. A proper implementation would have to loop through it to find the item to replace. It should usually be the first one, but AFAIK there are no guarantees.

This implementation is nice and simple and should be fine for now. If we ever change `Target` to have mutable fields, it can be refactored. But I don't think this needs to come with this change. It's fine as it is.